crypto/engine/eng_lib.c

```diff
@@ -71,10 +71,14 @@ int engine_free_util(ENGINE *e, int locked)
 
     if (e == NULL)
         return 1;
-    if (locked)
+#ifdef HAVE_ATOMICS
+    CRYPTO_DOWN_REF(&e->struct_ref, &i, global_engine_lock);
+#else
+    if (!locked)
```

Comment:
On Mon, Aug 29, 2016 at 05:22:27PM -0700, jrchamp wrote:
> If HAVE_ATOMICS is not defined, this represents a reversal of the logic from `if (locked)` to `if (!locked)`; was this intended?

Yes, the old code seems to have a bug.  If it's already locked
there is no reason to lock it again.